app/src/main/java/io/gnosis/safe/utils/MnemonicAddressDerivation.kt

```diff
@@ -0,0 +1,53 @@
+package io.gnosis.safe.utils
+
+import okio.ByteString
+import pm.gnosis.crypto.HDNode
+import pm.gnosis.crypto.KeyGenerator
+import pm.gnosis.mnemonic.Bip39
+import pm.gnosis.model.Solidity
+import pm.gnosis.utils.asBigInteger
+import java.math.BigInteger
+
+interface MnemonicAddressDerivation {
+    fun initialize(mnemonic: String)
+    fun addressesForRange(range: LongRange): List<Solidity.Address>
+    fun addressesForPage(start: Long, pageSize: Int): List<Solidity.Address>
+}
+
+interface MnemonicKeyDerivation {
+    fun initialize(mnemonic: String)
+    fun keyForIndex(index: Long): BigInteger
+}
+
+class MnemonicKeyAndAddressDerivation(
+        private val bip39 : Bip39
+) : MnemonicKeyDerivation, MnemonicAddressDerivation {
+
+    private lateinit var masterKey: HDNode
+
+    override fun initialize(mnemonic: String) {
+        val mnemonicSeed = bip39.mnemonicToSeed(mnemonic)
+        val hdNode = KeyGenerator.masterNode(ByteString.of(*mnemonicSeed))
+        masterKey = hdNode.derive(KeyGenerator.BIP44_PATH_ETHEREUM)
+    }
+
+    override fun addressesForRange(range: LongRange): List<Solidity.Address> {
+        val result = mutableListOf<Solidity.Address>()
+        (range).map {
```

Comment:
do we need to enclose range in parentheses?

Reply:
Changed in 164da5b29bba2da94c71e9fe5ca461c071398f90